File: src/ck3mm/workspace.py
"""Repository and per-mod workspace discovery."""

from __future__ import annotations

import tomllib
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Any

from .config import Config, ConfigError

ROOT_MARKER = "ck3mm.toml"
SOURCE_KINDS = frozenset({"workshop", "game", "repository", "mod"})
ARTIFACT_PREFIX = "artifacts"


class WorkspaceError(ValueError):
    """Raised when repository or mod workspace metadata is invalid."""


def discover_root(start: Path | None = None) -> Path:
    """Find the nearest ancestor containing the repository's ``ck3mm.toml``."""
    candidate = Path.cwd() if start is None else Path(start)
    if candidate.exists() and candidate.is_file():
        candidate = candidate.parent
    candidate = candidate.resolve(strict=False)
    for directory in (candidate, *candidate.parents):
        if (directory / ROOT_MARKER).is_file():
            return directory
    raise WorkspaceError(f"no {ROOT_MARKER} found from {candidate}")


def _read_toml(path: Path) -> dict[str, Any]:
    try:
        with path.open("rb") as handle:
            value = tomllib.load(handle)
    except FileNotFoundError as error:
        raise WorkspaceError(f"workspace metadata not found: {path}") from error
    except tomllib.TOMLDecodeError as error:
        raise WorkspaceError(f"invalid TOML in {path}: {error}") from error
    if not isinstance(value, dict):
        raise WorkspaceError(f"expected a TOML table in {path}")
    return value


def _relative_path(value: object, *, field_name: str, allow_glob: bool = False) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WorkspaceError(f"{field_name} must be a non-empty POSIX path")
    text = value.strip().replace("\\", "/")
    path = PurePosixPath(text)
    if path.is_absolute() or ".." in path.parts:
        raise WorkspaceError(
            f"{field_name} must stay within its declared root: {value}"
        )
    if not allow_glob and any(character in text for character in "*?["):
        raise WorkspaceError(f"{field_name} must not contain a glob: {value}")
    normalized = path.as_posix()
    if normalized in {"", "."}:
        raise WorkspaceError(f"{field_name} must not be empty")
    return normalized


def _string_list(value: object, *, field_name: str) -> tuple[str, ...]:
    if value is None:
        return ()
    if not isinstance(value, list) or not all(isinstance(item, str) for item in value):
        raise WorkspaceError(f"{field_name} must be an array of strings")
    return tuple(value)


@dataclass(frozen=True, slots=True)
class WorkspaceSettings:
    """Repository layout, loaded from ``ck3mm.toml`` at the workspace root."""

    mods_dir: str = "mods"
    tooling_dir: str = "workspace"
    state_dir: str = ".ignored/ck3mm"
    descriptor: str = "descriptor.mod"
    manifest: str = "mod.toml"
    artifacts_dir: str = ARTIFACT_PREFIX
    install_exclude: tuple[str, ...] = ("README.md",)

    @classmethod
    def from_root(cls, root: Path) -> WorkspaceSettings:
        data = _read_toml(root / ROOT_MARKER).get("workspace", {})
        if not isinstance(data, dict):
            raise WorkspaceError(f"{ROOT_MARKER}: workspace must be a table")
        defaults = cls()
        values: dict[str, Any] = {}
        for field_name in (
            "mods_dir",
            "tooling_dir",
            "state_dir",
            "descriptor",
            "manifest",
            "artifacts_dir",
        ):
            value = data.get(field_name, getattr(defaults, field_name))
            if not isinstance(value, str) or not value.strip():
                raise WorkspaceError(
                    f"{ROOT_MARKER}: workspace.{field_name} must be a non-empty string"
                )
            values[field_name] = value.strip()
        exclude = data.get("install_exclude")
        values["install_exclude"] = (
            _string_list(exclude, field_name="workspace.install_exclude")
            if exclude is not None
            else defaults.install_exclude
        )
        return cls(**values)


@dataclass(frozen=True, slots=True)
class SourceSpec:
    name: str
    kind: str
    path: str | None = None
    item_id: str | None = None
    mod: str | None = None

    @classmethod
    def from_mapping(cls, data: Mapping[str, Any], *, path: Path) -> SourceSpec:
        name = data.get("name")
        kind = data.get("kind")
        if not isinstance(name, str) or not name.strip():
            raise WorkspaceError(f"{path}: each source requires a non-empty name")
        if kind not in SOURCE_KINDS:
            raise WorkspaceError(
                f"{path}: source {name!r} has invalid kind {kind!r}; "
                f"expected one of {', '.join(sorted(SOURCE_KINDS))}"
            )

        source_path = data.get("path")
        normalized_path = (
            _relative_path(source_path, field_name=f"source {name}.path")
            if source_path not in (None, "")
            else None
        )
        item_id_value = data.get("item_id")
        item_id = str(item_id_value) if item_id_value is not None else None
        mod = data.get("mod")
        if mod is not None and (not isinstance(mod, str) or not mod.strip()):
            raise WorkspaceError(f"{path}: source {name}.mod must be a mod slug")

        if kind == "workshop" and not item_id:
            raise WorkspaceError(f"{path}: workshop source {name!r} requires item_id")
        if kind in {"game", "repository"} and not normalized_path:
            raise WorkspaceError(f"{path}: {kind} source {name!r} requires path")
        if kind == "mod" and not mod:
            raise WorkspaceError(f"{path}: mod source {name!r} requires mod")

        return cls(
            name=name.strip(),
            kind=kind,
            path=normalized_path,
            item_id=item_id,
            mod=mod.strip() if isinstance(mod, str) else None,
        )


@dataclass(frozen=True, slots=True)
class GeneratorSpec:
    entrypoint: str
    owned_outputs: tuple[str, ...]
    owned_artifacts: tuple[str, ...] = ()
    assets: tuple[str, ...] = ()

    @classmethod
    def from_mapping(cls, data: Mapping[str, Any], *, path: Path) -> GeneratorSpec:
        entrypoint = data.get("entrypoint")
        if not isinstance(entrypoint, str) or not entrypoint.strip():
            raise WorkspaceError(f"{path}: generator.entrypoint is required")
        module_path, separator, function = entrypoint.partition(":")
        if not separator or not function.isidentifier():
            raise WorkspaceError(
                f"{path}: generator.entrypoint must be FILE.py:function"
            )
        module_path = _relative_path(
            module_path, field_name="generator.entrypoint module"
        )
        if not module_path.endswith(".py"):
            raise WorkspaceError(f"{path}: generator entrypoint must use a .py file")

        outputs = tuple(
            _relative_path(item, field_name="generator.owned_outputs", allow_glob=True)
            for item in _string_list(
                data.get("owned_outputs"), field_name="generator.owned_outputs"
            )
        )
        if not outputs:
            raise WorkspaceError(f"{path}: generator.owned_outputs must not be empty")
        for output in outputs:
            if PurePosixPath(output).parts[0] == ARTIFACT_PREFIX:
                raise WorkspaceError(
                    f"{path}: {ARTIFACT_PREFIX}/ is reserved for "
                    f"generator.owned_artifacts: {output}"
                )

        artifacts = tuple(
            _relative_path(
                item, field_name="generator.owned_artifacts", allow_glob=True
            )
            for item in _string_list(
                data.get("owned_artifacts", []),
                field_name="generator.owned_artifacts",
            )
        )

        assets = tuple(
            _relative_path(item, field_name="generator.assets", allow_glob=True)
            for item in _string_list(
                data.get("assets", []), field_name="generator.assets"
            )
        )
        return cls(
            entrypoint=f"{module_path}:{function}",
            owned_outputs=outputs,
            owned_artifacts=artifacts,
            assets=assets,
        )


@dataclass(frozen=True, slots=True)
class ModManifest:
    path: Path
    slug: str
    generator: GeneratorSpec | None = None
    sources: tuple[SourceSpec, ...] = ()


@dataclass(frozen=True, slots=True)
class Mod:
    """One mod: an installable payload root plus its non-shipping tooling root."""

    slug: str
    root: Path
    tooling_root: Path
    descriptor_path: Path
    manifest_path: Path
    artifacts_root: Path
    manifest: ModManifest | None = None


def _source_mappings(value: object, *, path: Path) -> list[Mapping[str, Any]]:
    if value is None:
        return []
    if isinstance(value, list):
        if not all(isinstance(item, dict) for item in value):
            raise WorkspaceError(f"{path}: sources must contain tables")
        return value
    if isinstance(value, dict):
        result: list[Mapping[str, Any]] = []
        for name, entry in value.items():
            if not isinstance(entry, dict):
                raise WorkspaceError(f"{path}: source {name!r} must be a table")
            result.append({"name": name, **entry})
        return result
    raise WorkspaceError(f"{path}: sources must be an array of tables")


def load_manifest(path: Path, *, default_slug: str) -> ModManifest:
    """Load one ``workspace/<slug>/mod.toml`` manifest."""
    data = _read_toml(path)
    slug = default_slug.strip()
    if slug in {".", ".."} or "/" in slug or "\\" in slug:
        raise WorkspaceError(f"invalid mod directory name {slug!r}")

    generator_data = data.get("generator")
    if generator_data is not None and not isinstance(generator_data, dict):
        raise WorkspaceError(f"{path}: generator must be a table")
    generator = (
        GeneratorSpec.from_mapping(generator_data, path=path)
        if generator_data is not None
        else None
    )
    sources = tuple(
        SourceSpec.from_mapping(item, path=path)
        for item in _source_mappings(data.get("sources"), path=path)
    )
    source_names = [source.name for source in sources]
    if len(source_names) != len(set(source_names)):
        raise WorkspaceError(f"{path}: source names must be unique")

    return ModManifest(
        path=path.resolve(),
        slug=slug,
        generator=generator,
        sources=sources,
    )


@dataclass(frozen=True, slots=True)
class Workspace:
    root: Path
    settings: WorkspaceSettings

    @classmethod
    def from_path(cls, start: Path | None = None) -> Workspace:
        root = discover_root(start)
        return cls(root=root, settings=WorkspaceSettings.from_root(root))

    @property
    def mods_dir(self) -> Path:
        return self.root / self.settings.mods_dir

    @property
    def tooling_dir(self) -> Path:
        return self.root / self.settings.tooling_dir

    @property
    def state_dir(self) -> Path:
        return self.root / self.settings.state_dir

    def _mod_at(self, slug: str) -> Mod | None:
        mod_root = self.mods_dir / slug
        tooling_root = self.tooling_dir / slug
        descriptor = mod_root / self.settings.descriptor
        manifest_path = tooling_root / self.settings.manifest
        if not descriptor.is_file() and not manifest_path.is_file():
            return None
        manifest = (
            load_manifest(manifest_path, default_slug=slug)
            if manifest_path.is_file()
            else None
        )
        return Mod(
            slug=slug,
            root=mod_root,
            tooling_root=tooling_root,
            descriptor_path=descriptor,
            manifest_path=manifest_path,
            artifacts_root=tooling_root / self.settings.artifacts_dir,
            manifest=manifest,
        )

    def iter_mods(self) -> Iterator[Mod]:
        """Yield mods in stable slug order, keyed by their payload directory."""
        if not self.mods_dir.is_dir():
            return
        for slug in sorted(
            path.name for path in self.mods_dir.iterdir() if path.is_dir()
        ):
            mod = self._mod_at(slug)
            if mod is not None:
                yield mod

    def mods(self) -> tuple[Mod, ...]:
        return tuple(self.iter_mods())

    def get_mod(self, slug: str) -> Mod:
        for mod in self.iter_mods():
            if mod.slug == slug:
                return mod
        raise WorkspaceError(f"unknown local mod: {slug}")

    def manifests(self) -> tuple[ModManifest, ...]:
        return tuple(mod.manifest for mod in self.iter_mods() if mod.manifest)

    def resolve_source(
        self,
        source: SourceSpec,
        config: Config,
        *,
        must_exist: bool = True,
    ) -> Path:
        """Resolve a portable logical source into a local read-only path."""
        if source.kind == "workshop":
            if config.workshop_dir is None:
                raise ConfigError("CK3_WORKSHOP_DIR is required for Workshop sources")
            base = config.workshop_dir / str(source.item_id)
        elif source.kind == "game":
            if config.game_dir is None:
                raise ConfigError("CK3_GAME_DIR is required for game sources")
            base = config.game_dir
        elif source.kind == "repository":
            base = self.root
        elif source.kind == "mod":
            base = self.get_mod(source.mod or "").root
        else:  # pragma: no cover - SourceSpec validates this boundary.
            raise WorkspaceError(f"unsupported source kind: {source.kind}")

        resolved_base = base.resolve(strict=False)
        resolved = (resolved_base / (source.path or ".")).resolve(strict=False)
        if not resolved.is_relative_to(resolved_base):
            raise WorkspaceError(f"source escapes its declared root: {source.name}")
        if must_exist and not resolved.exists():
            raise WorkspaceError(f"source {source.name!r} does not exist: {resolved}")
        return resolved

    def resolve_sources(
        self,
        manifest: ModManifest,
        config: Config,
        *,
        must_exist: bool = True,
    ) -> dict[str, Path]:
        return {
            source.name: self.resolve_source(source, config, must_exist=must_exist)
            for source in manifest.sources
        }

File: internal/workspace/workspace.go
// Package workspace discovers the repository root and pairs each installable
// payload under mods/ with its non-shipping tooling under workspace/.
package workspace

import (
	"fmt"
	"os"
	"path"
	"path/filepath"
	"sort"
	"strings"

	"github.com/BurntSushi/toml"

	"codeberg.org/traidare/ck3_mods/internal/config"
	"codeberg.org/traidare/ck3_mods/internal/fsutil"
)

// RootMarker names the file whose presence marks the workspace root.
const RootMarker = "ck3mm.toml"

// ArtifactPrefix is the staging prefix reserved for generated development
// artifacts. Generators stage there; the runner promotes into workspace/.
const ArtifactPrefix = "artifacts"

// SourceKinds are the logical source roots a manifest may declare.
var SourceKinds = []string{"game", "mod", "repository", "workshop"}

// DiscoverRoot walks up from start to the nearest directory holding ck3mm.toml.
func DiscoverRoot(start string) (string, error) {
	candidate := start
	if candidate == "" {
		working, err := os.Getwd()
		if err != nil {
			return "", fmt.Errorf("cannot determine the working directory: %w", err)
		}
		candidate = working
	}
	if info, err := os.Stat(candidate); err == nil && !info.IsDir() {
		candidate = filepath.Dir(candidate)
	}
	candidate = fsutil.MustAbs(candidate)

	for {
		if fsutil.IsFile(filepath.Join(candidate, RootMarker)) {
			return candidate, nil
		}
		parent := filepath.Dir(candidate)
		if parent == candidate {
			return "", fmt.Errorf("no %s found from %s", RootMarker, fsutil.MustAbs(start))
		}
		candidate = parent
	}
}

// Settings is the repository layout declared by ck3mm.toml.
type Settings struct {
	ModsDir        string
	ToolingDir     string
	StateDir       string
	Descriptor     string
	Manifest       string
	ArtifactsDir   string
	InstallExclude []string
}

// DefaultSettings returns the layout assumed when ck3mm.toml omits a field.
func DefaultSettings() Settings {
	return Settings{
		ModsDir:        "mods",
		ToolingDir:     "workspace",
		StateDir:       ".ignored/ck3mm",
		Descriptor:     "descriptor.mod",
		Manifest:       "mod.toml",
		ArtifactsDir:   ArtifactPrefix,
		InstallExclude: []string{"README.md"},
	}
}

type settingsFile struct {
	Workspace struct {
		ModsDir        *string  `toml:"mods_dir"`
		ToolingDir     *string  `toml:"tooling_dir"`
		StateDir       *string  `toml:"state_dir"`
		Descriptor     *string  `toml:"descriptor"`
		Manifest       *string  `toml:"manifest"`
		ArtifactsDir   *string  `toml:"artifacts_dir"`
		InstallExclude []string `toml:"install_exclude"`
	} `toml:"workspace"`
}

// LoadSettings reads the [workspace] table from the root marker.
func LoadSettings(root string) (Settings, error) {
	markerPath := filepath.Join(root, RootMarker)
	var file settingsFile
	if _, err := toml.DecodeFile(markerPath, &file); err != nil {
		return Settings{}, fmt.Errorf("invalid TOML in %s: %w", markerPath, err)
	}

	settings := DefaultSettings()
	fields := []struct {
		name  string
		value *string
		into  *string
	}{
		{"mods_dir", file.Workspace.ModsDir, &settings.ModsDir},
		{"tooling_dir", file.Workspace.ToolingDir, &settings.ToolingDir},
		{"state_dir", file.Workspace.StateDir, &settings.StateDir},
		{"descriptor", file.Workspace.Descriptor, &settings.Descriptor},
		{"manifest", file.Workspace.Manifest, &settings.Manifest},
		{"artifacts_dir", file.Workspace.ArtifactsDir, &settings.ArtifactsDir},
	}
	for _, field := range fields {
		if field.value == nil {
			continue
		}
		trimmed := strings.TrimSpace(*field.value)
		if trimmed == "" {
			return Settings{}, fmt.Errorf("%s: workspace.%s must be a non-empty string", RootMarker, field.name)
		}
		*field.into = trimmed
	}
	if file.Workspace.InstallExclude != nil {
		settings.InstallExclude = file.Workspace.InstallExclude
	}
	return settings, nil
}

// SourceSpec is one logical, host-independent input a generator declares.
type SourceSpec struct {
	Name   string
	Kind   string
	Path   string
	ItemID string
	Mod    string
}

// GeneratorSpec declares a mod's generator and everything it owns.
//
// OwnedOutputs are payload paths promoted into mods/<slug>/; OwnedArtifacts are
// development paths promoted into workspace/<slug>/artifacts/ and never
// installed.
type GeneratorSpec struct {
	Entrypoint     string
	OwnedOutputs   []string
	OwnedArtifacts []string
}

// Manifest is one workspace/<slug>/mod.toml.
type Manifest struct {
	Path      string
	Slug      string
	Generator *GeneratorSpec
	Sources   []SourceSpec
}

// Mod pairs an installable payload root with its tooling root.
type Mod struct {
	Slug           string
	Root           string
	ToolingRoot    string
	DescriptorPath string
	ManifestPath   string
	ArtifactsRoot  string
	Manifest       *Manifest
}

// HasGenerator reports whether this mod's manifest declares a generator.
func (m Mod) HasGenerator() bool {
	return m.Manifest != nil && m.Manifest.Generator != nil
}

// relativePath validates a declared POSIX path that must stay inside its root.
func relativePath(value, fieldName string, allowGlob bool) (string, error) {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return "", fmt.Errorf("%s must be a non-empty POSIX path", fieldName)
	}
	text := strings.ReplaceAll(trimmed, `\`, "/")
	if strings.HasPrefix(text, "/") {
		return "", fmt.Errorf("%s must stay within its declared root: %s", fieldName, value)
	}
	for _, part := range strings.Split(text, "/") {
		if part == ".." {
			return "", fmt.Errorf("%s must stay within its declared root: %s", fieldName, value)
		}
	}
	if !allowGlob && strings.ContainsAny(text, "*?[") {
		return "", fmt.Errorf("%s must not contain a glob: %s", fieldName, value)
	}
	normalized := path.Clean(text)
	if normalized == "" || normalized == "." {
		return "", fmt.Errorf("%s must not be empty", fieldName)
	}
	return normalized, nil
}

type manifestFile struct {
	Generator *struct {
		Entrypoint     string   `toml:"entrypoint"`
		OwnedOutputs   []string `toml:"owned_outputs"`
		OwnedArtifacts []string `toml:"owned_artifacts"`
	} `toml:"generator"`
	Sources []struct {
		Name   string `toml:"name"`
		Kind   string `toml:"kind"`
		Path   string `toml:"path"`
		ItemID any    `toml:"item_id"`
		Mod    string `toml:"mod"`
	} `toml:"sources"`
}

func parseGenerator(raw manifestFile, manifestPath string) (*GeneratorSpec, error) {
	if raw.Generator == nil {
		return nil, nil
	}
	entrypoint := strings.TrimSpace(raw.Generator.Entrypoint)
	if entrypoint == "" {
		return nil, fmt.Errorf("%s: generator.entrypoint is required", manifestPath)
	}
	modulePath, function, found := strings.Cut(entrypoint, ":")
	if !found || !isIdentifier(function) {
		return nil, fmt.Errorf("%s: generator.entrypoint must be FILE.py:function", manifestPath)
	}
	modulePath, err := relativePath(modulePath, "generator.entrypoint module", false)
	if err != nil {
		return nil, err
	}
	if !strings.HasSuffix(modulePath, ".py") {
		return nil, fmt.Errorf("%s: generator entrypoint must use a .py file", manifestPath)
	}

	outputs, err := normalizeList(raw.Generator.OwnedOutputs, "generator.owned_outputs")
	if err != nil {
		return nil, err
	}
	if len(outputs) == 0 {
		return nil, fmt.Errorf("%s: generator.owned_outputs must not be empty", manifestPath)
	}
	for _, output := range outputs {
		if strings.Split(output, "/")[0] == ArtifactPrefix {
			return nil, fmt.Errorf("%s: %s/ is reserved for generator.owned_artifacts: %s",
				manifestPath, ArtifactPrefix, output)
		}
	}
	artifacts, err := normalizeList(raw.Generator.OwnedArtifacts, "generator.owned_artifacts")
	if err != nil {
		return nil, err
	}
	return &GeneratorSpec{
		Entrypoint:     modulePath + ":" + function,
		OwnedOutputs:   outputs,
		OwnedArtifacts: artifacts,
	}, nil
}

func normalizeList(values []string, fieldName string) ([]string, error) {
	result := make([]string, 0, len(values))
	for _, value := range values {
		normalized, err := relativePath(value, fieldName, true)
		if err != nil {
			return nil, err
		}
		result = append(result, normalized)
	}
	return result, nil
}

func parseSources(raw manifestFile, manifestPath string) ([]SourceSpec, error) {
	var sources []SourceSpec
	seen := map[string]bool{}
	for _, entry := range raw.Sources {
		name := strings.TrimSpace(entry.Name)
		if name == "" {
			return nil, fmt.Errorf("%s: each source requires a non-empty name", manifestPath)
		}
		if !validKind(entry.Kind) {
			return nil, fmt.Errorf("%s: source %q has invalid kind %q; expected one of %s",
				manifestPath, name, entry.Kind, strings.Join(SourceKinds, ", "))
		}

		source := SourceSpec{Name: name, Kind: entry.Kind, Mod: strings.TrimSpace(entry.Mod)}
		if strings.TrimSpace(entry.Path) != "" {
			normalized, err := relativePath(entry.Path, "source "+name+".path", false)
			if err != nil {
				return nil, err
			}
			source.Path = normalized
		}
		switch value := entry.ItemID.(type) {
		case nil:
		case string:
			source.ItemID = value
		case int64:
			source.ItemID = fmt.Sprintf("%d", value)
		default:
			source.ItemID = fmt.Sprintf("%v", value)
		}

		switch entry.Kind {
		case "workshop":
			if source.ItemID == "" {
				return nil, fmt.Errorf("%s: workshop source %q requires item_id", manifestPath, name)
			}
		case "game", "repository":
			if source.Path == "" {
				return nil, fmt.Errorf("%s: %s source %q requires path", manifestPath, entry.Kind, name)
			}
		case "mod":
			if source.Mod == "" {
				return nil, fmt.Errorf("%s: mod source %q requires mod", manifestPath, name)
			}
		}

		if seen[name] {
			return nil, fmt.Errorf("%s: source names must be unique", manifestPath)
		}
		seen[name] = true
		sources = append(sources, source)
	}
	return sources, nil
}

func validKind(kind string) bool {
	for _, candidate := range SourceKinds {
		if candidate == kind {
			return true
		}
	}
	return false
}

func isIdentifier(value string) bool {
	if value == "" {
		return false
	}
	for index, character := range value {
		isLetter := character == '_' ||
			(character >= 'a' && character <= 'z') ||
			(character >= 'A' && character <= 'Z')
		isDigit := character >= '0' && character <= '9'
		if !isLetter && !(index > 0 && isDigit) {
			return false
		}
	}
	return true
}

// LoadManifest reads one workspace/<slug>/mod.toml.
func LoadManifest(manifestPath, defaultSlug string) (*Manifest, error) {
	var raw manifestFile
	metadata, err := toml.DecodeFile(manifestPath, &raw)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("workspace metadata not found: %s", manifestPath)
		}
		return nil, fmt.Errorf("invalid TOML in %s: %w", manifestPath, err)
	}
	if undecoded := metadata.Undecoded(); len(undecoded) > 0 {
		keys := make([]string, len(undecoded))
		for index, key := range undecoded {
			keys[index] = key.String()
		}
		sort.Strings(keys)
		return nil, fmt.Errorf("unknown field(s) in %s: %s", manifestPath, strings.Join(keys, ", "))
	}
	slug := strings.TrimSpace(defaultSlug)
	if slug == "." || slug == ".." || strings.ContainsAny(slug, `/\`) {
		return nil, fmt.Errorf("invalid mod directory name %q", slug)
	}

	generator, err := parseGenerator(raw, manifestPath)
	if err != nil {
		return nil, err
	}
	sources, err := parseSources(raw, manifestPath)
	if err != nil {
		return nil, err
	}
	return &Manifest{
		Path:      fsutil.MustAbs(manifestPath),
		Slug:      slug,
		Generator: generator,
		Sources:   sources,
	}, nil
}

// Workspace is one repository checkout and its declared layout.
type Workspace struct {
	Root     string
	Settings Settings
}

// Open discovers the workspace root from start and loads its settings.
func Open(start string) (*Workspace, error) {
	root, err := DiscoverRoot(start)
	if err != nil {
		return nil, err
	}
	settings, err := LoadSettings(root)
	if err != nil {
		return nil, err
	}
	return &Workspace{Root: root, Settings: settings}, nil
}

// ModsDir is the installable payload tree.
func (w *Workspace) ModsDir() string { return filepath.Join(w.Root, w.Settings.ModsDir) }

// ToolingDir is the never-installed per-mod tooling tree.
func (w *Workspace) ToolingDir() string { return filepath.Join(w.Root, w.Settings.ToolingDir) }

// StateDir holds local, git-ignored tool state.
func (w *Workspace) StateDir() string { return filepath.Join(w.Root, w.Settings.StateDir) }

func (w *Workspace) modAt(slug string) (*Mod, error) {
	modRoot := filepath.Join(w.ModsDir(), slug)
	toolingRoot := filepath.Join(w.ToolingDir(), slug)
	descriptorPath := filepath.Join(modRoot, w.Settings.Descriptor)
	manifestPath := filepath.Join(toolingRoot, w.Settings.Manifest)

	hasDescriptor := fsutil.IsFile(descriptorPath)
	hasManifest := fsutil.IsFile(manifestPath)
	if !hasDescriptor && !hasManifest {
		return nil, nil
	}

	var manifest *Manifest
	if hasManifest {
		loaded, err := LoadManifest(manifestPath, slug)
		if err != nil {
			return nil, err
		}
		manifest = loaded
	}
	return &Mod{
		Slug:           slug,
		Root:           modRoot,
		ToolingRoot:    toolingRoot,
		DescriptorPath: descriptorPath,
		ManifestPath:   manifestPath,
		ArtifactsRoot:  filepath.Join(toolingRoot, w.Settings.ArtifactsDir),
		Manifest:       manifest,
	}, nil
}

// Mods returns every mod in stable slug order, keyed by its payload directory.
func (w *Workspace) Mods() ([]*Mod, error) {
	entries, err := os.ReadDir(w.ModsDir())
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	slugs := make([]string, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() {
			slugs = append(slugs, entry.Name())
		}
	}
	sort.Strings(slugs)

	var mods []*Mod
	for _, slug := range slugs {
		mod, err := w.modAt(slug)
		if err != nil {
			return nil, err
		}
		if mod != nil {
			mods = append(mods, mod)
		}
	}
	return mods, nil
}

// Mod returns one mod by slug.
func (w *Workspace) Mod(slug string) (*Mod, error) {
	mods, err := w.Mods()
	if err != nil {
		return nil, err
	}
	for _, mod := range mods {
		if mod.Slug == slug {
			return mod, nil
		}
	}
	return nil, fmt.Errorf("unknown local mod: %s", slug)
}

// ResolveSource turns a portable logical source into a local read-only path.
func (w *Workspace) ResolveSource(source SourceSpec, settings config.Config, mustExist bool) (string, error) {
	var base string
	switch source.Kind {
	case "workshop":
		if settings.WorkshopDir == "" {
			return "", fmt.Errorf("CK3_WORKSHOP_DIR is required for Workshop sources")
		}
		base = filepath.Join(settings.WorkshopDir, source.ItemID)
	case "game":
		if settings.GameDir == "" {
			return "", fmt.Errorf("CK3_GAME_DIR is required for game sources")
		}
		base = settings.GameDir
	case "repository":
		base = w.Root
	case "mod":
		mod, err := w.Mod(source.Mod)
		if err != nil {
			return "", err
		}
		base = mod.Root
	default:
		return "", fmt.Errorf("unsupported source kind: %s", source.Kind)
	}

	resolvedBase := fsutil.MustAbs(base)
	relative := source.Path
	if relative == "" {
		relative = "."
	}
	resolved := fsutil.MustAbs(filepath.Join(resolvedBase, filepath.FromSlash(relative)))
	if _, inside := fsutil.RelativeWithin(resolvedBase, resolved); !inside {
		return "", fmt.Errorf("source escapes its declared root: %s", source.Name)
	}
	if mustExist {
		if _, err := os.Stat(resolved); err != nil {
			return "", fmt.Errorf("source %q does not exist: %s", source.Name, resolved)
		}
	}
	return resolved, nil
}

// ResolveSources resolves every source a manifest declares, by name.
func (w *Workspace) ResolveSources(manifest *Manifest, settings config.Config, mustExist bool) (map[string]string, error) {
	resolved := map[string]string{}
	for _, source := range manifest.Sources {
		path, err := w.ResolveSource(source, settings, mustExist)
		if err != nil {
			return nil, err
		}
		resolved[source.Name] = path
	}
	return resolved, nil
}
